doc: extend README explaining how version bumping should be handled (#78)
zephyrproject-rtos/mbedtls: version-bumping.md

Scope of this document
======================

> Explain how to update the Zephyr's fork of Mbed TLS to align with a newer

Reviewer: ```suggestion
Explain how to update Zephyr's fork of Mbed TLS to align with a newer
```

upstream release.
This document talks about Mbed TLS, but the very same applies also to
TF-PSA-Crypto that, starting from Mbed TLS release 4.0, will be part of the

> Mbed TLS project.

Reviewer: We also want this to apply to any TF-M fork, so this statement could be made more generic in that regard.

Author: Yeah but this document is part of the Mbed TLS repo, not the main Zephyr one. In the latter case I would have no objection in mentioning different modules, but since we're in the Mbed TLS module repo I don't think it has much sense to mention other modules.

===================================

- Working branch name: `development`
- 1 or more LTS branches named as `mbedtls-<major>.<minor>` (ex: `mbedtls-3.6`)
- Tag at every release with format `v<major>.<minor>.<patch>` (ex: `v3.6.4`)

How was done in the past?
=========================

Zephyr's fork of Mbed TLS contains some extra patches that are not applied (and
that won't be applied) upstream.

So the procedure for each version bump was as follows:
- revert all the Zephyr's additional patches;
- merge all the changes from upstream Mbed TLS;
- re-apply all Zephyr's additional patches.

Graphically prior to each release we had:

```
* Mbed TLS commit #1
* Mbed TLS commit #2
* Mbed TLS commit #3 ("v3.6.3" tag)
* Zephyr commit #1
* Zephyr commit #2
* Zephyr commit #3 ("zephyr" branch HEAD)
```

and then after it:

```
* Mbed TLS commit #1
* Mbed TLS commit #2
* Mbed TLS commit #3 ("v3.6.3" tag)
* Zephyr commit #1
* Zephyr commit #2
* Zephyr commit #3
* Revert Zephyr commit #3
* Revert Zephyr commit #2
* Revert Zephyr commit #1
* Mbed TLS commit #4
* Mbed TLS commit #5
* Mbed TLS commit #6 ("v3.6.4" tag)
* Zephyr commit #1
* Zephyr commit #2
* Zephyr commit #3 ("zephyr" branch HEAD)

> ```

Reviewer: Do we really want to document the old way? We might if this is an RFC, but if it's meant as a reference document IMO no.

Author: Agree. This also simplifies the document :)

Proposed approach
=================

Create a branch named `zephyr-development` that includes:

- all Mbed TLS' commits on `development` branch up to a certain tag and
- Zephyr's custom commits.

Graphically:

```
* Mbed TLS commit #1
* Mbed TLS commit #2
* Mbed TLS commit #3 ("v3.6.3" tag)
 \
  * Zephyr commit #1
  * Zephyr commit #2
  * Zephyr commit #3 ("zephyr-development" branch HEAD)
```

All new Zephyr's patches go on top of `zephyr-development` HEAD.

When a new Mbed TLS release is announced then:

> - current Zephyr's HEAD commit is tagged adding a `zephyr_` prefix to the

Reviewer: I guess you mean the HEAD of `zephyr-development` here

>   upstream Mbed TLS' tag;

Reviewer: ```suggestion
  upstream Mbed TLS tag;
```

>   upstream Mbed TLS' tag;

Reviewer: So for example if we're updating to 3.6.5 from 3.6.4 the tag would be `zephyr_v3.6.4`?
Might be a bit confusing, how about `zephyr_mbedtls-<upstream tag>` to make sure we don't confuse Zephyr and Mbed TLS versions? For example right now we have a branch for Zephyr 3.7 LTS in the TF-M repo called [v3.7-branch](https://github.com/zephyrproject-rtos/trusted-firmware-m/tree/v3.7-branch).

> - rebase on the upstream tag of the new release.

Reviewer: When do you do the rebase? As part of the PR or beforehand?

Graphically:

```
* Mbed TLS commit #1
* Mbed TLS commit #2
* Mbed TLS commit #3 (Mbed TLS tag #1 - ex: "v3.6.3")
|\
| * Zephyr commit #1
| * Zephyr commit #2
| * Zephyr commit #3 ("zephyr-v3.6.3" tag)
|
* Mbed TLS commit #4
* Mbed TLS commit #5

> * Mbed TLS commit #6 (Mbed TLS tag #1 - ex: v3.6.4)

Reviewer: ```suggestion
* Mbed TLS commit #6 (Mbed TLS tag #2 - ex: "v3.6.4")
```

Scope of this document
======================

Explain how to update Zephyr's fork of Mbed TLS to align with a newer upstream
release.
This document talks about Mbed TLS, but the very same should apply also to

> TF-PSA-Crypto (which starting from Mbed TLS release 4.0 will be part of the

Reviewer: I don't think this kind of detail is relevant here?

Reviewer: You now dropped the first sentence:
> Explain how to update Zephyr's fork of Mbed TLS to align with a newer upstream release.

But considering this section's title is `Scope of this document` I think it makes sense to have such an introductory phrase?
Anyway if this is combined with the patches guidelines into the readme it can be reworked later.

> Mbed TLS project), TF-M and other modules.

Reviewer: We could simplify this to something like
```suggestion
This document talks about Mbed TLS, but the same guidelines apply to
the other forks of the TF-M and Mbed TLS projects.
```

Upstream Mbed TLS branches and tags
===================================

- Working branch name: `development`

> - 1 or more LTS branches named as `mbedtls-<major>.<minor>` (ex: `mbedtls-3.6`)

Reviewer: Why **LTS** branches? we want to cover any release, not just LTS ones

Author: Yes, the 1st line talks about the `development` branch and the 2nd one about LTS-es. I assume this was not clear enough, so I will rephrase it a bit.

Reviewer: Non-LTS Mbed TLS versions never have an associated release branch (e.g. for fix releases)?

- Tag at every release with format `v<major>.<minor>.<patch>` (ex: `v3.6.4`)

Proposed approach
=================

Let's say that a new Mbed TLS release is announced and that its release tag
is for example `v3.6.5`.
On the Zephyr fork:

> 1. Create a new branch named `zephyr-mbedtls-v3.6.5` which is located exactly at

Reviewer: To follow Zephyr's generic module guidelines where it's not burdening, maybe let's use `zephyr_` rather than `zephyr-`: https://docs.zephyrproject.org/latest/develop/modules.html#:~:text=the%20%E2%80%98zephyr_%E2%80%99%20prefix

Author: I will also update the target merge branch of #79

>    the upstream release tag.

Reviewer: The thing is that other projects (TF-M) have different branch and tag naming conventions, so I think we should make up generic guidelines that can be applied identically to both without having to wonder.
For example:
`Create a new branch named zephyr_<project name>-v<upstream release version> [...] For example, <include Mbed TLS example here>.`

> 2. Make a PR by cherry-picking all the Zephyr extra patches that are still valid

Reviewer: I would say `needed` rather than `valid`.

> 2. Make a PR by cherry-picking all the Zephyr extra patches that are still valid

Reviewer: Let's maybe make it explicit to not cherry pick them with `-x` nor `-s` to keep the patches as-is (and not inflate their commit messages over time)?

   for the new branch [1]. For this PR set the target merge branch to the newly

>    created `zephyr-mbedtls-v3.6.5`. [2]

Reviewer: The [1] and [2] are not formatted: https://github.com/valeriosetti/zephyr-mbedtls/blob/upstream-sync-strategy/version-bumping.md
Maybe it's possible to make them actual footnotes?

> 3. Wait for the PR to be merged.

Reviewer: ```suggestion
3. Get the PR approved and merged.
```

> 4. Set the repo `default` branch to `zephyr-mbedtls-v3.6.5`.

Reviewer: What do we do with the previous default branch (here that would be `zephyr-mbedtls-v3.6.4`)? I would be in favor of deleting it if there's no need to keep it alive. In that case before doing that it would need to be tagged. So some instructions should be added for this.

> [1]: Likely all the `[zep fromtree]` (and perhaps `[zep fromlist]`) should be

Reviewer: I wouldn't necessarily say "likely". It depends on what release we're updating to, what the upstream project is doing, what cherry picks we have...
I would rather just say that whatever fromtrees and fromlists are already included in the new release must be discarded.

     discarded as they should already be included in the new upstream release.
     Others (i.e. `[zep noup]`) might no more be valid or might require an

>      update. In general each Zephyr extra patch should be checked individually.

Reviewer: Let's also instruct to reorder the commits to follow the order `fromtree` - `fromlist` - `noup` (if/when applicable)?

Scope of this document
======================

This document talks about Mbed TLS, but the same guidelines apply to
the other forks of the TF-M and Mbed TLS projects.

Upstream Mbed TLS branches and tags
===================================

- 1 working branch name named `development`, where the ongoing work happens.
- 1 or more LTS branches named as `mbedtls-<major>.<minor>` (ex: `mbedtls-3.6`).

> - Tag at every release with format `v<major>.<minor>.<patch>` (ex: `v3.6.4`).

Reviewer: I don't think this is very relevant, considering we rather wanna make the guidelines generic enough so they apply just the same to the other projects e.g. TF-M.
We can just say that every project has releases and associated tags and sometimes dedicated release branches? or just entirely omit this section

Proposed approach
=================

Let's say that a new Mbed TLS release is announced and that its release tag
is for example `v3.6.5`.
On the Zephyr fork:
1. Create a new branch named `zephyr_<project_name>_v<upstream_release_version>`

>    which is located exactly at the upstream release tag. For example

Reviewer: ```suggestion
   which is located exactly at the upstream release tag. For example,
```

   `zephyr_mbedtls_v3.6.5`.
2. Make a PR by cherry-picking all the Zephyr extra patches that are still needed

>    for the new branch [1][2].

Reviewer: ```suggestion
   for the new branch.
```

>    - Discard all the `[zep fromtree]`and `[zep fromlist]` that are already

Reviewer: ```suggestion
   - Discard all the `[zep fromtree]` and `[zep fromlist]` commits that are already
```

      included in the release. As for `[zep noup]` they might no more be valid
      or might require an update so they should be checked individually.
   - The cherry-picking order should be: `[zep fromtree]` first, then
       `[zep fromlist]` and finally `[zep noup]`.
   - Cherry-pick commits without `-x` nor `-s` so as not to inflate the commit

>      message over time.

Reviewer: Caveat:
```suggestion
   - Cherry-pick commits without `-x` nor `-s` so as not to inflate the commit
     message over time, unless you had to modify the commit and it's not just a clean cherry-pick.
```

> 3. For the created PR set the target merge branch to the newly created

Reviewer: GitHub terminology:
```suggestion
3. For the created PR set the base branch to the newly-created
```

>    `zephyr_mbedtls_v3.6.5`.

Reviewer: ```suggestion
   one. In our example, it would be `zephyr_mbedtls_v3.6.5`.
```

>    - With this approach the PR in GitHub will show only Zephyr extra commits

Reviewer: ```suggestion
   - With this approach the PR in GitHub will only show Zephyr's extra commits
```

     on top of the upstream release tag, so it will be easier to review.

> 4. Get the PR approved and merged.

Reviewer: Obviously here there's the implicit step of making a PR to update the West revision in the Zephyr repo but I guess we don't need to explicitly say it?

5. Set the repo default branch to `zephyr_mbedtls_v3.6.5`.
   - When this is done, previous default branch (which in our example would be
     `zephyr_mbedtls_v3.6.4`) is removed from the repo and replaced with a tag

>      in order to keep its history.

Reviewer: What syntax for the tags? I checked now and it's possible to push a tag named the same as a branch, so we could go for that?
\+
```suggestion
   - When this is done, the HEAD of the previous default branch (which in our example would be
     `zephyr_mbedtls_v3.6.4`) is tagged (in order to keep its history) with the name same as the branch and the branch is deleted.
```
